### PR TITLE
Try-state for Referenda pallet

### PR DESCRIPTION
**Pull request by Szegoo**
Part of: https://github.com/paritytech/polkadot-sdk/issues/239

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2698716

File: frame/referenda/src/lib.rs [hunk 1]
```diff
@@ -1270,4 +1280,140 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+				ReferendumInfoFor::<T, I>::iter_keys().count(),
+			DispatchError::Other(
+				"Number of referenda in `ReferendumInfoFor` is greater than `ReferendumCount`"
+			)
+		);
+
+		// The order of indices when running `queueing_works` is: 0, 4, 2, 1, 3 🤔 
+		/*
+		ensure!(
+			ReferendumInfoFor::<T, I>::iter_keys()
```
Comment (Szegoo) [inline frame/referenda/src/lib.rs:1328]: Any idea why is this happening?

File: frame/referenda/src/lib.rs [hunk 2]
```diff
@@ -1270,4 +1280,140 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+				// Timed out proposals cannot be in the deciding phase.
+				ensure!(
+					(TrackQueue::<T, I>::get(track.0).iter().count()).saturating_sub(timed_out) == 0,
+					DispatchError::Other("`TrackQueue` should be empty when `DecidingCount` is less than `TrackInfo::max_decidin`.")
+				);
+			}
+
+			TrackQueue::<T, I>::get(track.0).iter().try_for_each(|(referendum_index, _)| -> DispatchResult {
+				ensure!(
+					ReferendumInfoFor::<T, I>::contains_key(referendum_index),
```
Comment (Szegoo) [inline frame/referenda/src/lib.rs:1409]: Cargo fmt struggles with formatting this line.

**Comment by Szegoo** [comment]
@ggwpez Could you take a quick look?

File: frame/referenda/src/lib.rs [hunk 3]
```diff
@@ -1270,4 +1280,140 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	///
+	/// * `ReferendumCount` must always be equal to the number of referenda in `ReferendumInfoFor`.
+	/// * Referendum indices must be sorted in the `ReferendumInfoOf` storage map.
+	/// * Referendum indices in `MetadataOf` must also be stored in `ReferendumInfoOf`.
+	///
+	/// Looking at referenda info:
+	///
+	/// - Data regarding ongoing phase:
+	///
+	/// * The submission deposit cannot be less than `T::SubmissionDeposit`.
```
Comment (muharem) [inline frame/referenda/src/lib.rs:1298]: it can be, if it changed.

File: frame/referenda/src/lib.rs [hunk 4]
```diff
@@ -1270,4 +1280,140 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	/// * There must exist track info for the track of the referendum.
+	/// * The decision deposit if provided cannot be less than the decision deposit of the track.
```
Comment (muharem) [inline frame/referenda/src/lib.rs:1300]: same, it can be if changed

File: frame/referenda/src/lib.rs [hunk 5]
```diff
@@ -1270,4 +1280,140 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	/// * The deciding stage has to begin before confirmation period.
```
Comment (muharem) [inline frame/referenda/src/lib.rs:1301]: if there is enough support and approves accumulated during preparation period, referendum will move into confirmation stage skipping the decision stage.

File: frame/referenda/src/lib.rs [hunk 6]
```diff
@@ -1270,4 +1280,140 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	///
+	/// - Data in `ReferendumInfo::Approved`, `ReferendumInfo::Rejected`,
+	///   `ReferendumInfo::Cancelled` and `ReferendumInfo::TimedOut`:
+	///
+	/// * The submission deposit cannot be less than `T::SubmissionDeposit`.
```
Comment (muharem) [inline frame/referenda/src/lib.rs:1306]: duplication here

File: frame/referenda/src/lib.rs [hunk 7]
```diff
@@ -1270,4 +1280,140 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	///
+	/// Looking at tracks:
+	/// * The `TrackQueue` should be empty if `DecidingCount` is less than
```
Comment (muharem) [inline frame/referenda/src/lib.rs:1309]: I think this is possible when something is removed from the decision stage, and only with the next block the task from the queue will be removed.

File: frame/referenda/src/lib.rs [hunk 8]
```diff
@@ -1270,4 +1280,140 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	///   `TrackInfo::max_deciding`.
+	/// * The referendum indices stored in TrackQueue` must exist as keys in the `ReferendumInfoFor`
```
Comment (muharem) [inline frame/referenda/src/lib.rs:1311]:
```suggestion
	/// * The referendum indices stored in `TrackQueue` must exist as keys in the `ReferendumInfoFor`
```

File: frame/referenda/src/lib.rs (hunk 1 above)
Comment (muharem) [reply frame/referenda/src/lib.rs:1328]: probably this does not return in order it was added, the StorageMap is a tree

File: frame/referenda/src/migration.rs [hunk 9]
```diff
@@ -82,6 +82,10 @@ pub mod v0 {
 	#[storage_alias]
 	pub type ReferendumInfoFor<T: Config<I>, I: 'static> =
 		StorageMap<Pallet<T, I>, Blake2_128Concat, ReferendumIndex, ReferendumInfoOf<T, I>>;
+
+	#[storage_alias]
+	pub type ReferendumCount<T: Config<I>, I: 'static> =
```
Comment (muharem) [inline frame/referenda/src/migration.rs:87]: this item not used, below you use the storage from the pallet.

Comment (Szegoo) [reply frame/referenda/src/migration.rs:87]: Yes, I initially used this but forgot to remove it later.

**Comment by Szegoo** [comment]
@muharem Thanks for your review, I am going to move this into draft now, I want to give this some more thought to see if something should be added/removed.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3040521

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3048548

File: frame/referenda/src/lib.rs [hunk 10]
```diff
@@ -1283,4 +1296,102 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
 		}
 	}
+
+	/// Ensure the correctness of the state of this pallet.
+	///
+	/// The following assertions must always apply.
+	///
+	/// General assertions:
+	///
+	/// * `ReferendumCount` must always be equal to the number of referenda in `ReferendumInfoFor`.
```
Comment (kianenigma) [inline frame/referenda/src/lib.rs:1306]: Why is this not a counted map?

File: frame/referenda/src/lib.rs (hunk 7 above)
Comment (kianenigma) [reply frame/referenda/src/lib.rs:1320]: If in doubt, please remove it form the PR so this can be moved forward.

**Review by kianenigma — APPROVED** [review]
LGTM assuming that the existing assumptions all pass against Poladot and Kusama. A second pair of eyes by @gavofyork or @muharem would be great.

**Comment by Szegoo** [comment]
> LGTM assuming that the existing assumptions all pass against Poladot and Kusama.

I will run the checks against the Polkadot and Kusama storage and will share the results here.

File: frame/referenda/src/lib.rs [hunk 11]
```diff
@@ -1283,4 +1296,84 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	/// - Data regarding ongoing phase:
+	///
+	/// * There must exist track info for the track of the referendum.
+	/// * The deciding stage has to begin before confirmation period.
+	/// * If alarm is set the nudge call has to be at most `UndecidingTimeout` blocks away
+	///  from the submission block.
+	///
+	/// Looking at tracks:
+	/// * The referendum indices stored in `TrackQueue` must exist as keys in the
+	///   `ReferendumInfoFor`
```
Comment (gpestana) [inline frame/referenda/src/lib.rs:1321]: fmt

File: frame/referenda/src/lib.rs [hunk 12]
```diff
@@ -1283,4 +1296,84 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	///  storage map.
+	#[cfg(any(feature = "try-runtime", test))]
+	fn do_try_state() -> Result<(), TryRuntimeError> {
+		ensure!(
+			ReferendumCount::<T, I>::get() as usize ==
+				ReferendumInfoFor::<T, I>::iter_keys().count(),
+			"Number of referenda in `ReferendumInfoFor` is greater than `ReferendumCount`"
```
Comment (gpestana) [inline frame/referenda/src/lib.rs:1328]:
```suggestion
			"Number of referenda in `ReferendumInfoFor` is different than `ReferendumCount`"
```

**Review by gpestana — COMMENTED** [review]
I usually find that breaking down the logic in `try_state` separately is easier to read and understand what invariants are being checked. Especially because you often return a `DispatchResult` from each check closure within `do_try_runtime`. 

An example of breaking down the try runtime checks is implemented in the [EPM pallet](https://github.com/paritytech/substrate/blob/master/frame/election-provider-multi-phase/src/lib.rs#L1583). IMO this is not a deal breaker, but just thought of putting it out there for comparison and for you to decide which approach is better.

Thanks for helping here!

**Comment by Szegoo** [comment]
@gpestana I agree it does look nice when separated into multiple functions.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3081607

File: frame/referenda/src/lib.rs [hunk 13]
```diff
@@ -1283,4 +1296,90 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+						ensure!(
+							deciding.since <
+								deciding.confirming.unwrap_or(T::BlockNumber::max_value()),
+								"Deciding status cannot begin before confirming stage."
+						)
+					}
+
+					if let Some(alarm) = status.alarm {
+						ensure!(alarm.0 <= status.submitted.saturating_add(T::UndecidingTimeout::get()), "The alarm cannot be set more than `UndecidingTimeout` away from the submission block.")
+					}
```
Comment (Szegoo) [inline frame/referenda/src/lib.rs:1358]:
```suggestion
```
This is true for the first alarm, but the alarm can be set after the first nudge call(e.g. at the beginning of the deciding period). For this reason, this check has been failing against the Polkadot state and I am removing it.

**Comment by Szegoo** [comment]
Screenshot 2023-06-30 17_30_47

CC @kianenigma

**Comment by liamaharon** [comment]
@Szegoo could you also please check westend and rococo?

**Comment by Szegoo** [comment]
@liamaharon AFAIU they don't implement the referenda pallet.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3285439

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3285440

**Comment by liamaharon** [comment]
can you please rebase?

File: frame/referenda/src/migration.rs [hunk 14]
```diff
@@ -188,6 +188,14 @@ pub mod test {
 		}
 	}
 
+	fn increment_referendum_index() {
+		ReferendumCount::<T, ()>::mutate(|x| {
+			let r = *x;
+			*x += 1;
+			r
+		});
+	}
```
Comment (liamaharon) [inline frame/referenda/src/migration.rs:197]: I think you can just do

```suggestion
	fn increment_referendum_index() {
        ReferendumCount::<T, ()>::mutate(|x| x.saturating_inc());
    }
```

so probably don't need `increment_referendum_index` and can do it inline

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3285644

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3285629

File: frame/referenda/src/mock.rs [hunk 15]
```diff
@@ -225,23 +225,32 @@ impl Config for Test {
...
+    ext.execute_with(|| System::set_block_number(1));
+    ext
+	}
+
+	pub fn build_and_execute(self, test: impl FnOnce() -> ()) {
+		self.build().execute_with(|| {
+			test();
+			Referenda::do_try_state().unwrap();
+		})
+	}
```
Comment (liamaharon) [inline frame/referenda/src/mock.rs]: formatting looks off here?

File: frame/referenda/src/lib.rs [hunk 16]
```diff
@@ -1284,4 +1294,86 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
 	}
+
+	/// Ensure the correctness of the state of this pallet.
+	///
+	/// The following assertions must always apply.
+	///
+	/// General assertions:
+	///
+	/// * `ReferendumCount` must always be equal to the number of referenda in `ReferendumInfoFor`.
+	/// * Referendum indices in `MetadataOf` must also be stored in `ReferendumInfoFor`.
```
Comment (liamaharon) [inline frame/referenda/src/lib.rs:1305]:
```suggestion
	/// * Referendum indices in [`MetadataOf`] must also be stored in [`ReferendumInfoFor`].
```

Comment (liamaharon) [reply frame/referenda/src/lib.rs:1305]: and in other comments please

File: frame/referenda/src/lib.rs [hunk 17]
```diff
@@ -1284,4 +1294,87 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	/// * Referendum indices in [`MetadataOf`] must also be stored in [`ReferendumInfoFor`].
+	#[cfg(any(feature = "try-runtime", test))]
+	fn do_try_state() -> Result<(), sp_runtime::TryRuntimeError> {
+		ensure!(
+			ReferendumCount::<T, I>::get() as usize ==
+				ReferendumInfoFor::<T, I>::iter_keys().count(),
+			"Number of referenda in `ReferendumInfoFor` is different than `ReferendumCount`"
+		);
+
+		MetadataOf::<T, I>::iter_keys().try_for_each(|referendum_index| -> DispatchResult {
```
Comment (gpestana) [inline frame/referenda/src/lib.rs:1315]: Should we also check whether `ReferendumInfoFor` has metadata keys lingering? A count equality check would suffice after this look (ie. `ReferendumInfoFor::count == MetadataOf:::count`)

**Comment by gpestana** [comment]
bot fmt

**Review by gpestana — APPROVED** [review]
Left a comment but lgtm, thanks for the PR!

**Review by liamaharon — APPROVED** [review]

File: frame/referenda/src/lib.rs (hunk 17 above)
Comment (Szegoo) [reply frame/referenda/src/lib.rs:1315]: `ReferendumInfoFor::count` won't always equal to `MetadataOf:::count`. The metadata is only set when calling the `set_metadata` extrinsic

**Comment by liamaharon** [comment]
bot merge

**Comment by liamaharon** [comment]
@paritytech/tip-bot-approvers could you please tip?

**Comment by juangirini** [comment]
@Szegoo could you please add your dot/ksm address to the PR description for a tip
format "[Polkadot|Kusama] address: xxx"

**Comment by juangirini** [comment]
@rzadp that's awesome, thanks for the update

**Pull request merged**